src/main/java/org/kuali/kfs/krad/service/impl/AttachmentServiceImpl.java

```diff
@@ -0,0 +1,290 @@
+        FileOutputStream streamOut = null;
+        BufferedOutputStream bufferedStreamOut = null;
+        try {
+            streamOut = new FileOutputStream(fileOut);
+            bufferedStreamOut = new BufferedOutputStream(streamOut);
+            int c;
+            while ((c = fileContents.read()) != -1) {
+                bufferedStreamOut.write(c);
+            }
+        } finally {
+            bufferedStreamOut.close();
+            streamOut.close();
+        }
+    }
+
+    public void moveAttachmentWherePending(Note note) {
+        if (note == null) {
+            throw new IllegalArgumentException("Note must be non-null");
+        }
+        if (StringUtils.isBlank(note.getObjectId())) {
+            throw new IllegalArgumentException("Note does not have a valid object id, object id was null or empty");
+        }
+        Attachment attachment = note.getAttachment();
+        if (attachment != null) {
+            try {
+                moveAttachmentFromPending(attachment, note.getRemoteObjectIdentifier());
+            } catch (IOException e) {
+                throw new RuntimeException("Problem moving pending attachment to final directory");
+            }
+        }
+    }
+
+    private void moveAttachmentFromPending(Attachment attachment, String objectId) throws IOException {
+        //This method would probably be more efficient if attachments had a pending flag
+        String fullPendingFileName = getPendingDirectory() + File.separator + attachment.getAttachmentIdentifier();
+        File pendingFile = new File(fullPendingFileName);
+
+        if (pendingFile.exists()) {
+            BufferedInputStream bufferedStream = null;
+            FileInputStream oldFileStream = null;
+            String fullPathNewFile = getDocumentDirectory(objectId) + File.separator + attachment.getAttachmentIdentifier();
+            try {
+                oldFileStream = new FileInputStream(pendingFile);
+                bufferedStream = new BufferedInputStream(oldFileStream);
+                writeInputStreamToFileStorage(bufferedStream, fullPathNewFile);
+            } finally {
+
+                bufferedStream.close();
+                oldFileStream.close();
+                //this has to come after the close
+                pendingFile.delete();
+
+            }
+        }
+
+    }
+
+    public void deleteAttachmentContents(Attachment attachment) {
+        if (attachment.getNote() == null) throw new RuntimeException("Attachment.note must be set in order to delete the attachment");
+        String fullPathUniqueFileName = getDocumentDirectory(attachment.getNote().getRemoteObjectIdentifier()) + File.separator + attachment.getAttachmentIdentifier();
+        File attachmentFile = new File(fullPathUniqueFileName);
+        attachmentFile.delete();
+    }
+
+    private String getPendingDirectory() {
+        return this.getDocumentDirectory("");
+    }
+
+    protected String getDocumentDirectory(String objectId) {
```

Comment:
The only change in this class is the visibility of thus method, can we create a contribution for that we can remove this in the future?